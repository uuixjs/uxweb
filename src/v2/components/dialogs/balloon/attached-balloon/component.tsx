import { FC } from "react";
import { Display, Layout } from "../../../layout";
import { Attached, AttachedProps } from "../../attached/component";
import {
  BalloonWrapper,
  BalloonWrapperProps,
} from "../balloon-wrapper/component";

export interface AttachedBalloonProps
  extends BalloonWrapperProps,
    AttachedProps {
  /**
   * @example true
   */
  show?: boolean;
}

/**
 * Legacy implementation of Balloon which uses CSS position: absolute
 *
 * @deprecated Please migrate to the new `<Balloon>` component; see Core UI 19.0.0 release notes and docs for migration steps.
 */
export const AttachedBalloon: FC<AttachedBalloonProps> = ({
  show,
  direction,
  offsetX,
  offsetY,
  ...props
}) => (
  // Display.Hide was the legacy way this worked; keeping for compatibility
  <Layout display={show ? Display.Block : Display.Hide}>
    <Attached direction={direction} offsetX={offsetX} offsetY={offsetY}>
      <BalloonWrapper {...props} />
    </Attached>
  </Layout>
);

AttachedBalloon.displayName = "AttachedBalloon";
