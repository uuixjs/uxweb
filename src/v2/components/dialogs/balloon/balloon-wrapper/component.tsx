import { Background, Color, Elevation, InjectLayout } from "../../../layout";
import {
  BorderRadius,
  CSSObject,
  css,
  staticTokenRule,
  styled,
  themeTokenRule,
} from "lib/ui-utils";
import { FC, ReactNode } from "react";

import { withOverlayContext } from "../../../overlay-region/context";

export enum BalloonSize {
  /* Auto means the balloon width should adjust to accomodate the width of its children without wrapping */
  Auto = "auto",
  ExtraSmall = "xs",
  Small = "sm",
  Medium = "md",
  Large = "lg",
  ExtraLarge = "xl",
}

const BalloonSizeElevations: { [key: string]: Elevation } = {
  [BalloonSize.Auto]: 2,
  [BalloonSize.ExtraSmall]: 2,
  [BalloonSize.Small]: 2,
  [BalloonSize.Medium]: 3,
  [BalloonSize.Large]: 4,
  [BalloonSize.ExtraLarge]: 5,
};

const BalloonSizes: { [key: string]: CSSObject } = {
  [BalloonSize.Auto]: {
    "min-width": 0,
    "white-space": "nowrap",
  },
  [BalloonSize.ExtraSmall]: { width: "10rem", "min-width": "10rem" },
  [BalloonSize.Small]: { width: "20rem" },
  [BalloonSize.Medium]: { width: "30rem" },
  [BalloonSize.Large]: { width: "40rem" },
  [BalloonSize.ExtraLarge]: { width: "50rem" },
};

export interface BalloonWrapperProps {
  /** Set the background color of the balloon */
  background?: Background;
  /**
   * Content rendered in the Balloon
   *
   * @example <Layout padding={5} />
   */
  children?: ReactNode;
  /** Apply elevation styling to the balloon. */
  elevation?: Elevation;
  /**
   * The width of the Balloon. A Balloon's height is determined by
   * its children.
   */
  size?: BalloonSize | "auto" | "xs" | "sm" | "md" | "lg" | "xl"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;
  /** Adds overlay support to Balloon. */
  overlay?: boolean;
  /**
   * Passes an `aria-label` to the component to provide context to assistive
   * devices. All standard HTML aria attributes are now supported on Core UI
   * components â€” see https://www.w3.org/TR/html-aria/."
   */
  "aria-label"?: string;
}

const ScBalloonWrapper = styled.div<{
  $size: BalloonWrapperProps["size"];
  $overlay: BalloonWrapperProps["overlay"];
}>`
  display: inline-block;
  min-width: 16rem;
  max-width: 90vw;
  ${(props) => {
    return props.$size ? BalloonSizes[props.$size] : undefined;
  }};

  ${(props) => {
    if (props.$overlay) {
      return css`
        border-style: solid;
        border-width: ${staticTokenRule("border-width-default")};
        border-color: ${themeTokenRule("color-border-balloon-overlay")};
      `;
    }
  }};
`;

export const DEFAULT_BALLOON_ELEVATION = 2;

/**
 * To facilitate a focused interface, important actions and information should
 * always be visible to the user; but, less-used actions and information can
 * be placed within Balloons and shown by clicking.
 */
export const BalloonWrapperComponent: FC<BalloonWrapperProps> = ({
  background,
  elevation,
  size,
  overlay,
  ...props
}) => {
  return (
    <InjectLayout
      borderRadius={BorderRadius.Large}
      background={overlay ? Background.Overlay : background}
      elevation={
        elevation === 0
          ? elevation // when the elevation is specified as 0, set elevation 0.
          : size && elevation
          ? elevation // when elevation and size is otherwise specified, use that elevation regardless of size
          : (size && BalloonSizeElevations[size]) || // if size is specified, but not elevation, use the elevation according to size
            elevation || // without a specified size, but with elevation, use the elevation specified
            DEFAULT_BALLOON_ELEVATION // otherwise, use the default elevation
      }
      color={overlay ? Color.Overlay : Color.Inherit}
    >
      <ScBalloonWrapper
        $size={size}
        $overlay={overlay}
        role="dialog"
        className="tw-balloon"
        {...props}
      />
    </InjectLayout>
  );
};

BalloonWrapperComponent.defaultProps = {
  background: Background.Base,
};

BalloonWrapperComponent.displayName = "BalloonWrapperComponent";

export const BalloonWrapper = withOverlayContext(BalloonWrapperComponent);
