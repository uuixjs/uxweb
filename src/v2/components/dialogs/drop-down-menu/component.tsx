import { DialogLayerProps } from "../dialog-layer";
import { withDialogLayer } from "../dialog-layer/with-dialog-layer";
import {
  DropDownMenuWrapper,
  DropDownMenuWrapperProps,
} from "./drop-down-menu-wrapper";

export interface DropDownMenuProps
  extends DropDownMenuWrapperProps,
    DialogLayerProps {}

export const DropDownMenu = withDialogLayer(DropDownMenuWrapper);
