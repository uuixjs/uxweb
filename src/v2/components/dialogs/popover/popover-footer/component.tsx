import { AlignItems, Display, JustifyContent, Layout } from "../../../layout";
import { BorderRadius, getAriaProps, getDataProps } from "lib/ui-utils";
import { Button, ButtonProps, ButtonType } from "../../../button/button";

import { Component } from "react";

export interface PopoverFooterProps {
  primaryButtonProps: ButtonProps;
  secondaryButtonProps?: ButtonProps;
}

export class PopoverFooter extends Component<PopoverFooterProps> {
  public render() {
    return (
      <Layout
        {...getDataProps(this.props)}
        {...getAriaProps(this.props)}
        padding={1}
        display={Display.Flex}
        alignItems={AlignItems.Center}
        justifyContent={JustifyContent.End}
        borderRadius={{
          bottomLeft: BorderRadius.Medium,
          bottomRight: BorderRadius.Medium,
        }}
      >
        {this.props.secondaryButtonProps && (
          <Layout padding={{ right: 1 }}>
            <Button
              variant={ButtonType.Secondary}
              {...this.props.secondaryButtonProps}
            />
          </Layout>
        )}
        <Button
          variant={ButtonType.Primary}
          {...this.props.primaryButtonProps}
        />
      </Layout>
    );
  }
}
