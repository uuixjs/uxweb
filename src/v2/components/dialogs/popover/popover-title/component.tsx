import { FC } from "react";
import { CoreText, TextType } from "../../../core-text";
import { Color, FontWeight, InjectLayout, TextAlign } from "../../../layout";
import { withOverlayContext } from "../../../overlay-region/context";

export interface PopoverTitleProps {
  title: string;
  subtitle?: string;
  overlay?: boolean;
}

export const PopoverTitleComponent: FC<PopoverTitleProps> = (props) => {
  const { title, subtitle } = props;
  return (
    <>
      <InjectLayout
        textAlign={TextAlign.Center}
        fontWeight={FontWeight.SemiBold}
      >
        <CoreText
          type={TextType.H5}
          color={props.overlay ? Color.Overlay : Color.Alt}
        >
          {title}
        </CoreText>
      </InjectLayout>
      {subtitle ? (
        <InjectLayout textAlign={TextAlign.Center}>
          <CoreText
            type={TextType.H6}
            color={props.overlay ? Color.OverlayAlt : Color.Alt2}
          >
            {subtitle}
          </CoreText>
        </InjectLayout>
      ) : null}
    </>
  );
};

PopoverTitleComponent.displayName = "PopoverTitle";
export const PopoverTitle = withOverlayContext(PopoverTitleComponent);
