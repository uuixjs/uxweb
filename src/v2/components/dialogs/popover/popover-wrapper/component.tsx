import {
  Background,
  Display,
  Elevation,
  FlexDirection,
  InjectLayout,
} from "../../../layout";
import { BalloonSize, BalloonWrapper } from "../../balloon/balloon-wrapper";
import { Component, ReactNode } from "react";
import { getAriaProps, getDataProps } from "lib";

export interface PopoverWrapperProps {
  /** This ID attribute must be the same for all elements for a given popover in order to set the correct aria attribute values. In other words, a Popover, PopoverHeader and PopoverBody should each receive the same string. */
  popoverId: string;
  elevation?: Elevation;
  /**
   * Sets a maximum height to the menu using CSS units (`rem` is preferred).
   * Contents will scroll if taller than the maximum height.
   *
   * @example 20rem
   */
  maxHeight?: string;
  size?: BalloonSize;
  children?: ReactNode;
  background?: Background;
  /** Render without positioning instead of absolute. This is a temporary prop to ease migration toward a future change to this component. */
  inline?: boolean;
}

export class PopoverWrapper extends Component<PopoverWrapperProps, {}> {
  public render() {
    return (
      <BalloonWrapper
        {...getDataProps(this.props)}
        {...getAriaProps(this.props)}
        elevation={this.props.elevation}
        size={this.props.size ? this.props.size : BalloonSize.Large}
        background={this.props.background}
      >
        <InjectLayout
          display={Display.Flex}
          flexDirection={FlexDirection.Column}
          role="dialog"
          aria-labelledby={`${this.props.popoverId}-header`}
          aria-describedby={`${this.props.popoverId}-body`}
        >
          <div style={{ maxHeight: this.props.maxHeight }}>
            {this.props.children}
          </div>
        </InjectLayout>
      </BalloonWrapper>
    );
  }
}
