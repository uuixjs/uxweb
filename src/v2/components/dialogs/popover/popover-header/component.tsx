import {
  AlignItems,
  Color,
  Display,
  FlexDirection,
  InjectLayout,
  JustifyContent,
  Layout,
} from "../../../layout";
import {
  AttachedTooltip,
  TooltipAlign,
  TooltipDirection,
} from "../../tooltip/attached-tooltip";
import {
  BorderRadius,
  MarginProps,
  getMarginStyles,
  styled,
} from "lib/ui-utils";
import {
  ButtonIcon,
  ButtonIconProps,
  ButtonIconType,
} from "../../../button/button-icon";
import { CloseButtonProps, CoreDismissible } from "../../../core-dismissible";
import { FC, ReactNode } from "react";
import { PopoverTitle, PopoverTitleProps } from "../popover-title";

import { withOverlayContext } from "../../../overlay-region/context";

export interface PopoverHeaderProps {
  /** This ID attribute must be the same for all elements for a given popover in order to set the correct aria attribute values. In other words, a Popover, PopoverHeader and PopoverBody should each receive the same string. */
  popoverId: string;
  title?: PopoverTitleProps | ReactNode;
  buttonLeftPrimary?: ButtonIconProps;
  buttonLeftSecondary?: ButtonIconProps;
  buttonRightPrimary: CloseButtonProps;
  buttonRightSecondary?: ButtonIconProps;
  overlay?: boolean;
}

const ScPopoverHeader = styled(Layout)`
  min-height: 5rem;
  display: flex;
  flex-shrink: 0;
`;

const ScPopoverHeaderIconSlot = styled.div<MarginProps>`
  ${getMarginStyles}
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 3rem;
`;

export const PopoverHeaderComponent: FC<PopoverHeaderProps> = ({
  buttonLeftPrimary,
  buttonLeftSecondary,
  buttonRightPrimary,
  buttonRightSecondary,
  overlay,
  popoverId,
  title,
  ...props
}) => {
  const leftPrimaryButton = buttonLeftPrimary || buttonLeftSecondary;
  const leftSecondaryButton = buttonLeftPrimary && buttonLeftSecondary;
  const rightPrimaryButton = buttonRightPrimary || buttonRightSecondary;
  const rightSecondaryButton = buttonRightPrimary && buttonRightSecondary;

  return (
    <ScPopoverHeader
      className="tw-popover-header"
      color={overlay ? Color.Overlay : Color.Base}
      borderRadius={{
        topLeft: BorderRadius.Medium,
        topRight: BorderRadius.Medium,
      }}
      elevation={1}
      padding={{ y: 0.5, x: 1 }}
      {...props}
    >
      {leftPrimaryButton || rightPrimaryButton ? (
        <ScPopoverHeaderIconSlot
          margin={{ right: 0.5 }}
          className="tw-popover-header__icon-slot--left"
        >
          {renderButton(leftPrimaryButton)}
        </ScPopoverHeaderIconSlot>
      ) : null}
      {leftSecondaryButton || rightSecondaryButton ? (
        <ScPopoverHeaderIconSlot
          margin={{ right: 0.5 }}
          className="tw-popover-header__icon-slot--left"
        >
          {renderButton(leftSecondaryButton)}
        </ScPopoverHeaderIconSlot>
      ) : null}
      <InjectLayout
        flexGrow={1}
        justifyContent={JustifyContent.Center}
        alignItems={AlignItems.Center}
        flexDirection={FlexDirection.Column}
        display={Display.Flex}
      >
        <div id={`${popoverId}-header`}>
          {isPopoverHeaderTitleProps(title) ? (
            <PopoverTitle title={title.title} subtitle={title.subtitle} />
          ) : (
            title
          )}
        </div>
      </InjectLayout>
      {leftSecondaryButton || rightSecondaryButton ? (
        <ScPopoverHeaderIconSlot
          margin={{ left: 0.5 }}
          className="tw-popover-header__icon-slot--right"
        >
          {renderButton(rightSecondaryButton)}
        </ScPopoverHeaderIconSlot>
      ) : null}
      {leftPrimaryButton || rightPrimaryButton ? (
        <ScPopoverHeaderIconSlot
          margin={{ left: 0.5 }}
          className="tw-popover-header__icon-slot--right"
        >
          {renderButton(rightPrimaryButton, true)}
        </ScPopoverHeaderIconSlot>
      ) : null}
    </ScPopoverHeader>
  );
};

function renderButton(
  props: ButtonIconProps | CloseButtonProps | undefined,
  closeButton: boolean = false,
) {
  if (!props) {
    return null;
  }
  const button = closeButton ? (
    <CoreDismissible
      {...(props as CloseButtonProps)}
      variant={ButtonIconType.Secondary}
      title={undefined}
    />
  ) : (
    <ButtonIcon
      {...(props as ButtonIconProps)}
      variant={ButtonIconType.Secondary}
      title={undefined}
    />
  );
  return props.title ? (
    <AttachedTooltip
      label={props.title}
      align={TooltipAlign.Center}
      direction={TooltipDirection.Bottom}
    >
      {button}
    </AttachedTooltip>
  ) : (
    button
  );
}

function isPopoverHeaderTitleProps(
  titleProps: PopoverTitleProps | ReactNode,
): titleProps is PopoverTitleProps {
  return !!(titleProps && titleProps.hasOwnProperty("title"));
}

PopoverHeaderComponent.displayName = "PopoverHeader";
export const PopoverHeader = withOverlayContext(PopoverHeaderComponent);
