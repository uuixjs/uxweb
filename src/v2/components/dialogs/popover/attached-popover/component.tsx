import { FC } from "react";
import { Display, Layout } from "../../../layout";
import { Attached, AttachedProps, BalloonDirection } from "../../attached";
import { PopoverWrapper, PopoverWrapperProps } from "../popover-wrapper";

export type PopoverDirection =
  | BalloonDirection.TopLeft
  | BalloonDirection.TopRight
  | BalloonDirection.BottomLeft
  | BalloonDirection.BottomRight;

export interface AttachedPopoverProps
  extends PopoverWrapperProps,
    AttachedProps {
  /** Direction the `Popover` balloon will go. */
  direction?: PopoverDirection;
  /**
   * Show the `Popover`. The `Popover`'s behavior is controlled by
   * its parent.
   */
  show?: boolean;
}

/**
 * Legacy implementation of Popover which uses CSS position: absolute
 *
 * @deprecated Please migrate to the new `<Popover>` component; see Core UI 19.0.0 release notes and docs for migration steps.
 */
export const AttachedPopover: FC<AttachedPopoverProps> = ({
  show,
  direction,
  offsetX,
  offsetY,
  ...props
}) => (
  // Display.Hide was the legacy way this worked; keeping for compatibility
  <Layout display={show ? Display.Block : Display.Hide}>
    <Attached direction={direction} offsetX={offsetX} offsetY={offsetY}>
      <PopoverWrapper {...props} />
    </Attached>
  </Layout>
);
