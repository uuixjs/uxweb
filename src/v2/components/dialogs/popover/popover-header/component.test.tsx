import { mount } from "enzyme";
import { setupShallowTest } from "../../../../tests/helpers";
import {
  ButtonIcon,
  ButtonIconProps,
  ButtonIconType,
} from "../../../button/button-icon";
import { SVGAsset } from "../../../svg";
import { AttachedTooltip } from "../../tooltip/attached-tooltip";
import { PopoverHeaderComponent, PopoverHeaderProps } from "./component";

const BUTTON_ICON_PROPS = {
  "aria-label": "aria label",
  icon: SVGAsset.Gear,
};

function getRequiredProps(): PopoverHeaderProps {
  return {
    popoverId: "test-popover-id",
    buttonRightPrimary: {
      "aria-label": "Close",
    },
  };
}

function getPartialProps(): PopoverHeaderProps {
  return {
    popoverId: "test-popover-id",
    title: {
      title: "Title",
      subTitle: "SubTitle",
    },
    buttonRightPrimary: {
      "aria-label": "Close",
    },
    buttonLeftSecondary: {
      "aria-label": "label left secondary",
      icon: SVGAsset.Check,
    },
    buttonRightSecondary: {
      "aria-label": "label right secondary",
      icon: SVGAsset.Close,
    },
  };
}

function getOptionalProps(): PopoverHeaderProps {
  return {
    popoverId: "test-popover-id",
    title: {
      title: "Title",
      subTitle: "SubTitle",
    },
    buttonLeftPrimary: {
      "aria-label": "label left primary",
      icon: SVGAsset.GlyphArrLeft,
      variant: ButtonIconType.Alert,
    },
    buttonLeftSecondary: {
      "aria-label": "label left secondary",
      icon: SVGAsset.Check,
      variant: ButtonIconType.Secondary,
    },
    buttonRightPrimary: {
      "aria-label": "Close",
    },
    buttonRightSecondary: {
      "aria-label": "label right secondary",
      icon: SVGAsset.Close,
      variant: ButtonIconType.Primary,
    },
  };
}

function getOptionalPropsWithTooltips(): PopoverHeaderProps {
  return {
    popoverId: "test-popover-id",
    title: {
      title: "Title",
      subTitle: "SubTitle",
    },
    buttonLeftPrimary: {
      "aria-label": "label left primary",
      icon: SVGAsset.GlyphArrLeft,
      variant: ButtonIconType.Alert,
      title: "Alert",
    },
    buttonLeftSecondary: {
      "aria-label": "label left secondary",
      icon: SVGAsset.Check,
      variant: ButtonIconType.Secondary,
      title: "Mark All as Read",
    },
    buttonRightPrimary: {
      "aria-label": "Close",
      title: "Close Menu",
    },
    buttonRightSecondary: {
      "aria-label": "label right secondary",
      icon: SVGAsset.Close,
      variant: ButtonIconType.Primary,
      title: "Close",
    },
  };
}

function getChildrenProps(): PopoverHeaderProps {
  return {
    popoverId: "test-popover-id",
    title: "Title",
    buttonLeftPrimary: {
      "aria-label": "label left primary",
      icon: SVGAsset.GlyphArrLeft,
    },
    buttonLeftSecondary: {
      "aria-label": "label left secondary",
      icon: SVGAsset.Check,
    },
    buttonRightPrimary: {
      "aria-label": "Close",
    },
    buttonRightSecondary: {
      "aria-label": "label right secondary",
      icon: SVGAsset.Close,
    },
  };
}

const setupShallow = setupShallowTest(PopoverHeaderComponent, getRequiredProps);

describe("PopoverHeader", () => {
  it("renders Popover header component", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders Popover header component with secondary icons in the primary slot if they are the only ones provided", () => {
    const { wrapper } = setupShallow(getPartialProps());
    expect(wrapper).toMatchSnapshot();
  });

  it("renders Popover header component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper.find(AttachedTooltip)).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders Popover header component with all props and tooltips", () => {
    const { wrapper } = setupShallow(getOptionalPropsWithTooltips());
    expect(wrapper.find(AttachedTooltip)).toHaveLength(4);
  });

  it("renders Popover header component with child node instead of title/subtitle pattern", () => {
    const { wrapper } = setupShallow(getChildrenProps());
    expect(wrapper).toMatchSnapshot();
  });

  it("does not allow button type to be overwritten", () => {
    const BUTTON_PROPS: ButtonIconProps = {
      "aria-label": "aria label",
      variant: ButtonIconType.Primary,
      icon: SVGAsset.Gear,
    };

    const props = {
      ...getRequiredProps(),
      buttonLeftPrimary: BUTTON_PROPS,
      buttonLeftSecondary: BUTTON_PROPS,
      buttonRightPrimary: { "aria-label": "Close" },
      buttonRightSecondary: BUTTON_PROPS,
    };

    const wrapper = mount(<PopoverHeaderComponent {...props} />);

    expect(wrapper.find(ButtonIcon).at(0).prop("variant")).toBe(
      ButtonIconType.Secondary,
    );
    expect(wrapper.find(ButtonIcon).at(1).prop("variant")).toBe(
      ButtonIconType.Secondary,
    );
    expect(wrapper.find(ButtonIcon).at(2).prop("variant")).toBe(
      ButtonIconType.Secondary,
    );
    expect(wrapper.find(ButtonIcon).at(3).prop("variant")).toBe(
      ButtonIconType.Secondary,
    );
  });

  it("renders containers to balance the right side", () => {
    const { wrapper } = setupShallow({
      buttonLeftPrimary: BUTTON_ICON_PROPS,
    });

    expect(wrapper.find(".tw-popover-header__icon-slot--left").length).toBe(1);
    expect(wrapper.find(".tw-popover-header__icon-slot--right").length).toBe(1);
  });

  it("renders containers to balance the right side", () => {
    const { wrapper } = setupShallow({
      buttonLeftSecondary: BUTTON_ICON_PROPS,
    });

    expect(wrapper.find(".tw-popover-header__icon-slot--left").length).toBe(1);
    expect(wrapper.find(".tw-popover-header__icon-slot--right").length).toBe(1);
  });

  it("renders containers to balance the left side", () => {
    const { wrapper } = setupShallow({
      buttonRightPrimary: { "aria-label": "Close" },
    });

    expect(wrapper.find(".tw-popover-header__icon-slot--left").length).toBe(1);
    expect(wrapper.find(".tw-popover-header__icon-slot--right").length).toBe(1);
  });

  it("renders containers to balance the other side", () => {
    const { wrapper } = setupShallow({
      buttonLeftPrimary: BUTTON_ICON_PROPS,
      buttonLeftSecondary: BUTTON_ICON_PROPS,
    });

    expect(wrapper.find(".tw-popover-header__icon-slot--left").length).toBe(2);
    expect(wrapper.find(".tw-popover-header__icon-slot--right").length).toBe(2);
  });

  it("renders containers to balance the other side", () => {
    const { wrapper } = setupShallow({
      buttonRightPrimary: { "aria-label": "Close" },
      buttonRightSecondary: BUTTON_ICON_PROPS,
    });

    expect(wrapper.find(".tw-popover-header__icon-slot--left").length).toBe(2);
    expect(wrapper.find(".tw-popover-header__icon-slot--right").length).toBe(2);
  });

  it("does not render more than four containers", () => {
    const { wrapper } = setupShallow({
      buttonLeftPrimary: BUTTON_ICON_PROPS,
      buttonLeftSecondary: BUTTON_ICON_PROPS,
      buttonRightPrimary: { "aria-label": "Close" },
      buttonRightSecondary: BUTTON_ICON_PROPS,
    });

    expect(wrapper.find(".tw-popover-header__icon-slot--left").length).toBe(2);
    expect(wrapper.find(".tw-popover-header__icon-slot--right").length).toBe(2);
  });
});
