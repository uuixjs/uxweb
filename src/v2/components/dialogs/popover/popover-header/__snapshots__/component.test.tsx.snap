// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopoverHeader renders Popover header component 1`] = `
<Styled(Layout)
  borderRadius={
    Object {
      "topLeft": "border-radius-medium",
      "topRight": "border-radius-medium",
    }
  }
  className="tw-popover-header"
  color="base"
  elevation={1}
  padding={
    Object {
      "x": 1,
      "y": 0.5,
    }
  }
>
  <styled.div
    className="tw-popover-header__icon-slot--left"
    margin={
      Object {
        "right": 0.5,
      }
    }
  />
  <InjectLayout
    alignItems="center"
    display="flex"
    flexDirection="column"
    flexGrow={1}
    justifyContent="center"
  >
    <div
      id="test-popover-id-header"
    />
  </InjectLayout>
  <styled.div
    className="tw-popover-header__icon-slot--right"
    margin={
      Object {
        "left": 0.5,
      }
    }
  >
    <ForwardRef(CoreDismissible)
      aria-label="Close"
      variant="secondary"
    />
  </styled.div>
</Styled(Layout)>
`;

exports[`PopoverHeader renders Popover header component with all props 1`] = `
<Styled(Layout)
  borderRadius={
    Object {
      "topLeft": "border-radius-medium",
      "topRight": "border-radius-medium",
    }
  }
  className="tw-popover-header"
  color="base"
  elevation={1}
  padding={
    Object {
      "x": 1,
      "y": 0.5,
    }
  }
>
  <styled.div
    className="tw-popover-header__icon-slot--left"
    margin={
      Object {
        "right": 0.5,
      }
    }
  >
    <ButtonIcon
      aria-label="label left primary"
      icon="GlyphArrLeft"
      variant="secondary"
    />
  </styled.div>
  <styled.div
    className="tw-popover-header__icon-slot--left"
    margin={
      Object {
        "right": 0.5,
      }
    }
  >
    <ButtonIcon
      aria-label="label left secondary"
      icon="Check"
      variant="secondary"
    />
  </styled.div>
  <InjectLayout
    alignItems="center"
    display="flex"
    flexDirection="column"
    flexGrow={1}
    justifyContent="center"
  >
    <div
      id="test-popover-id-header"
    >
      <PopoverTitle
        title="Title"
      />
    </div>
  </InjectLayout>
  <styled.div
    className="tw-popover-header__icon-slot--right"
    margin={
      Object {
        "left": 0.5,
      }
    }
  >
    <ButtonIcon
      aria-label="label right secondary"
      icon="Close"
      variant="secondary"
    />
  </styled.div>
  <styled.div
    className="tw-popover-header__icon-slot--right"
    margin={
      Object {
        "left": 0.5,
      }
    }
  >
    <ForwardRef(CoreDismissible)
      aria-label="Close"
      variant="secondary"
    />
  </styled.div>
</Styled(Layout)>
`;

exports[`PopoverHeader renders Popover header component with child node instead of title/subtitle pattern 1`] = `
<Styled(Layout)
  borderRadius={
    Object {
      "topLeft": "border-radius-medium",
      "topRight": "border-radius-medium",
    }
  }
  className="tw-popover-header"
  color="base"
  elevation={1}
  padding={
    Object {
      "x": 1,
      "y": 0.5,
    }
  }
>
  <styled.div
    className="tw-popover-header__icon-slot--left"
    margin={
      Object {
        "right": 0.5,
      }
    }
  >
    <ButtonIcon
      aria-label="label left primary"
      icon="GlyphArrLeft"
      variant="secondary"
    />
  </styled.div>
  <styled.div
    className="tw-popover-header__icon-slot--left"
    margin={
      Object {
        "right": 0.5,
      }
    }
  >
    <ButtonIcon
      aria-label="label left secondary"
      icon="Check"
      variant="secondary"
    />
  </styled.div>
  <InjectLayout
    alignItems="center"
    display="flex"
    flexDirection="column"
    flexGrow={1}
    justifyContent="center"
  >
    <div
      id="test-popover-id-header"
    >
      Title
    </div>
  </InjectLayout>
  <styled.div
    className="tw-popover-header__icon-slot--right"
    margin={
      Object {
        "left": 0.5,
      }
    }
  >
    <ButtonIcon
      aria-label="label right secondary"
      icon="Close"
      variant="secondary"
    />
  </styled.div>
  <styled.div
    className="tw-popover-header__icon-slot--right"
    margin={
      Object {
        "left": 0.5,
      }
    }
  >
    <ForwardRef(CoreDismissible)
      aria-label="Close"
      variant="secondary"
    />
  </styled.div>
</Styled(Layout)>
`;

exports[`PopoverHeader renders Popover header component with secondary icons in the primary slot if they are the only ones provided 1`] = `
<Styled(Layout)
  borderRadius={
    Object {
      "topLeft": "border-radius-medium",
      "topRight": "border-radius-medium",
    }
  }
  className="tw-popover-header"
  color="base"
  elevation={1}
  padding={
    Object {
      "x": 1,
      "y": 0.5,
    }
  }
>
  <styled.div
    className="tw-popover-header__icon-slot--left"
    margin={
      Object {
        "right": 0.5,
      }
    }
  >
    <ButtonIcon
      aria-label="label left secondary"
      icon="Check"
      variant="secondary"
    />
  </styled.div>
  <styled.div
    className="tw-popover-header__icon-slot--left"
    margin={
      Object {
        "right": 0.5,
      }
    }
  />
  <InjectLayout
    alignItems="center"
    display="flex"
    flexDirection="column"
    flexGrow={1}
    justifyContent="center"
  >
    <div
      id="test-popover-id-header"
    >
      <PopoverTitle
        title="Title"
      />
    </div>
  </InjectLayout>
  <styled.div
    className="tw-popover-header__icon-slot--right"
    margin={
      Object {
        "left": 0.5,
      }
    }
  >
    <ButtonIcon
      aria-label="label right secondary"
      icon="Close"
      variant="secondary"
    />
  </styled.div>
  <styled.div
    className="tw-popover-header__icon-slot--right"
    margin={
      Object {
        "left": 0.5,
      }
    }
  >
    <ForwardRef(CoreDismissible)
      aria-label="Close"
      variant="secondary"
    />
  </styled.div>
</Styled(Layout)>
`;
